src/ray/gcs/store_client/redis_store_client.cc

```diff
@@ -214,7 +225,15 @@ Status RedisStoreClient::AsyncBatchDelete(const std::string &table_name,
   for (auto &key : keys) {
     redis_keys.push_back(GenRedisKey(external_storage_namespace_, table_name, key));
   }
-  return DeleteByKeys(redis_keys, callback);
+  return DeleteByKeys(redis_keys, [this, table_name, keys, callback](int64_t cnt) {
+    if (callback != nullptr) {
+      callback(cnt > 0);
+    }
+    absl::MutexLock lock(&cache_mu_);
```

Comment:
We need to delete (ditto for add) the keys from the cache *before* calling DeleteByKeys, and don't hold locks for it